File: server/internal/store/index_runs.go
package store

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
)

const (
	IndexRunStatusBuilding = "building"
	IndexRunStatusActive   = "active"
	IndexRunStatusRetired  = "retired"
	IndexRunStatusFailed   = "failed"
)

// IndexRun tracks one disposable index generation.
type IndexRun struct {
	ID             string         `json:"id"`
	TenantID       string         `json:"tenant_id,omitempty"`
	DomainID       string         `json:"domain_id,omitempty"`
	Backend        string         `json:"backend"`
	EmbeddingModel string         `json:"embedding_model"`
	ChunkSchema    map[string]any `json:"chunk_schema,omitempty"`
	Status         string         `json:"status"`
	ErrorMsg       string         `json:"error_msg,omitempty"`
	CreatedAt      string         `json:"created_at"`
	ActivatedAt    string         `json:"activated_at,omitempty"`
}

func scanIndexRun(row pgx.Row) (IndexRun, error) {
	var run IndexRun
	var tenantID, domainID, errMsg *string
	var chunkJSON []byte
	var createdAt time.Time
	var activatedAt *time.Time
	err := row.Scan(
		&run.ID, &tenantID, &domainID, &run.Backend, &run.EmbeddingModel, &chunkJSON,
		&run.Status, &errMsg, &createdAt, &activatedAt,
	)
	if err != nil {
		return run, err
	}
	if tenantID != nil {
		run.TenantID = *tenantID
	}
	if domainID != nil {
		run.DomainID = *domainID
	}
	if errMsg != nil {
		run.ErrorMsg = *errMsg
	}
	if len(chunkJSON) > 0 {
		_ = json.Unmarshal(chunkJSON, &run.ChunkSchema)
	}
	run.CreatedAt = createdAt.UTC().Format(time.RFC3339)
	if activatedAt != nil {
		run.ActivatedAt = activatedAt.UTC().Format(time.RFC3339)
	}
	return run, nil
}

// CreateIndexRun inserts a building index run.
func (st *ChatStore) CreateIndexRun(
	ctx context.Context,
	tenantID, domainID, backend, embeddingModel string,
	chunkSchema map[string]any,
) (IndexRun, error) {
	var zero IndexRun
	if backend == "" {
		backend = "chroma"
	}
	chunkJSON, err := json.Marshal(chunkSchema)
	if err != nil {
		return zero, err
	}
	id := uuid.NewString()
	var tPtr, dPtr *string
	if tenantID != "" {
		tPtr = &tenantID
	}
	if domainID != "" {
		dPtr = &domainID
	}
	row := st.Pool.QueryRow(ctx, `
		INSERT INTO index_runs (id, tenant_id, domain_id, backend, embedding_model, chunk_schema, status)
		VALUES ($1, $2, $3, $4, $5, $6::jsonb, $7)
		RETURNING id, tenant_id, domain_id, backend, embedding_model, chunk_schema, status, error_msg, created_at, activated_at`,
		id, tPtr, dPtr, backend, embeddingModel, chunkJSON, IndexRunStatusBuilding,
	)
	return scanIndexRun(row)
}

// ActivateIndexRun marks a run active and retires the previous active run for scope.
func (st *ChatStore) ActivateIndexRun(ctx context.Context, tenantID, domainID, runID string) error {
	tx, err := st.Pool.Begin(ctx)
	if err != nil {
		return err
	}
	defer func() { _ = tx.Rollback(ctx) }()

	_, err = tx.Exec(ctx, `
		UPDATE index_runs SET status = $1
		WHERE tenant_id IS NOT DISTINCT FROM NULLIF($2, '')::text
		  AND domain_id IS NOT DISTINCT FROM NULLIF($3, '')::text
		  AND status = $4 AND id <> $5`,
		IndexRunStatusRetired, tenantID, domainID, IndexRunStatusActive, runID,
	)
	if err != nil {
		return err
	}
	_, err = tx.Exec(ctx, `
		UPDATE index_runs SET status = $1, activated_at = NOW() WHERE id = $2`,
		IndexRunStatusActive, runID,
	)
	if err != nil {
		return err
	}
	_, err = tx.Exec(ctx, `
		INSERT INTO index_run_active (tenant_id, domain_id, index_run_id, updated_at)
		VALUES ($1, $2, $3, NOW())
		ON CONFLICT (tenant_id, domain_id) DO UPDATE SET index_run_id = EXCLUDED.index_run_id, updated_at = NOW()`,
		tenantID, domainID, runID,
	)
	if err != nil {
		return err
	}
	return tx.Commit(ctx)
}

// ActiveIndexRunID returns the active index run for tenant+domain scope.
func (st *ChatStore) ActiveIndexRunID(ctx context.Context, tenantID, domainID string) (string, error) {
	var runID string
	err := st.Pool.QueryRow(ctx, `
		SELECT index_run_id FROM index_run_active WHERE tenant_id = $1 AND domain_id = $2`,
		tenantID, domainID,
	).Scan(&runID)
	if errors.Is(err, pgx.ErrNoRows) {
		return "", nil
	}
	return runID, err
}

// UpsertIndexDocumentState records indexing outcome for a document in a run.
func (st *ChatStore) UpsertIndexDocumentState(
	ctx context.Context,
	runID, documentID string,
	indexedVersion int,
	contentSHA256 string,
	chunkCount int,
) error {
	_, err := st.Pool.Exec(ctx, `
		INSERT INTO index_document_state (index_run_id, document_id, indexed_version, content_sha256, chunk_count, indexed_at)
		VALUES ($1, $2, $3, $4, $5, NOW())
		ON CONFLICT (index_run_id, document_id) DO UPDATE SET
			indexed_version = EXCLUDED.indexed_version,
			content_sha256 = EXCLUDED.content_sha256,
			chunk_count = EXCLUDED.chunk_count,
			indexed_at = NOW()`,
		runID, documentID, indexedVersion, contentSHA256, chunkCount,
	)
	return err
}

// GetIndexRun loads an index run by id.
func (st *ChatStore) GetIndexRun(ctx context.Context, runID string) (*IndexRun, error) {
	row := st.Pool.QueryRow(ctx, `
		SELECT id, tenant_id, domain_id, backend, embedding_model, chunk_schema, status, error_msg, created_at, activated_at
		FROM index_runs WHERE id = $1`, runID)
	run, err := scanIndexRun(row)
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return &run, nil
}

// FinishIndexRun marks a run failed with message.
func (st *ChatStore) FinishIndexRun(ctx context.Context, runID, status, errMsg string) error {
	if status == "" {
		status = IndexRunStatusFailed
	}
	_, err := st.Pool.Exec(ctx, `
		UPDATE index_runs SET status = $1, error_msg = $2 WHERE id = $3`,
		status, nullIfEmpty(errMsg), runID,
	)
	return err
}

// DefaultChunkSchema returns the current chunking profile for index runs.
func DefaultChunkSchema() map[string]any {
	return map[string]any{
		"chunk_size":    500,
		"chunk_overlap": 50,
		"splitter":      "recursive",
		"schema":        1,
	}
}

// EnsureActiveIndexRun creates and activates an index run if none exists for scope.
func (st *ChatStore) EnsureActiveIndexRun(ctx context.Context, tenantID, domainID, backend, embeddingModel string) (string, error) {
	runID, err := st.ActiveIndexRunID(ctx, tenantID, domainID)
	if err != nil {
		return "", err
	}
	if runID != "" {
		return runID, nil
	}
	run, err := st.CreateIndexRun(ctx, tenantID, domainID, backend, embeddingModel, DefaultChunkSchema())
	if err != nil {
		return "", err
	}
	if err := st.ActivateIndexRun(ctx, tenantID, domainID, run.ID); err != nil {
		return "", fmt.Errorf("activate index run: %w", err)
	}
	return run.ID, nil
}

File: rag/kb/index_runs.py
"""Disposable index runs (blue/green index generations)."""

from __future__ import annotations

import json
import os
import uuid
from dataclasses import dataclass
from typing import Any

from rag.kb.documents import _connect
from rag.vector_backend.chroma_backend import EMBEDDING_MODEL, embedding_signature


@dataclass
class IndexRun:
    id: str
    tenant_id: str
    domain_id: str
    backend: str
    embedding_model: str
    chunk_schema: dict[str, Any]
    status: str


def default_chunk_schema() -> dict[str, Any]:
    return {"chunk_size": 500, "chunk_overlap": 50, "splitter": "recursive", "schema": 1}


def index_profile() -> dict[str, Any]:
    sig = embedding_signature()
    return {
        "backend": (os.environ.get("VECTOR_STORE") or "chroma").strip().lower(),
        "embedding_model": sig.get("model") or EMBEDDING_MODEL,
        "chunk_schema": default_chunk_schema(),
        "e5_prefixes": sig.get("e5_prefixes"),
        "schema": sig.get("schema", 1),
    }


def create_index_run(tenant_id: str, domain_id: str) -> str:
    profile = index_profile()
    run_id = str(uuid.uuid4())
    with _connect() as conn:
        with conn.cursor() as cur:
            cur.execute(
                """
                INSERT INTO index_runs (id, tenant_id, domain_id, backend, embedding_model, chunk_schema, status)
                VALUES (%s, NULLIF(%s, ''), NULLIF(%s, ''), %s, %s, %s::jsonb, 'building')
                """,
                (
                    run_id,
                    tenant_id,
                    domain_id,
                    profile["backend"],
                    profile["embedding_model"],
                    json.dumps(profile["chunk_schema"]),
                ),
            )
        conn.commit()
    return run_id


def activate_index_run(tenant_id: str, domain_id: str, run_id: str) -> None:
    with _connect() as conn:
        with conn.cursor() as cur:
            cur.execute(
                """
                UPDATE index_runs SET status = 'retired'
                WHERE tenant_id IS NOT DISTINCT FROM NULLIF(%s, '')::text
                  AND domain_id IS NOT DISTINCT FROM NULLIF(%s, '')::text
                  AND status = 'active' AND id <> %s
                """,
                (tenant_id, domain_id, run_id),
            )
            cur.execute(
                "UPDATE index_runs SET status = 'active', activated_at = NOW() WHERE id = %s",
                (run_id,),
            )
            cur.execute(
                """
                INSERT INTO index_run_active (tenant_id, domain_id, index_run_id, updated_at)
                VALUES (%s, %s, %s, NOW())
                ON CONFLICT (tenant_id, domain_id) DO UPDATE
                SET index_run_id = EXCLUDED.index_run_id, updated_at = NOW()
                """,
                (tenant_id, domain_id, run_id),
            )
        conn.commit()


def ensure_active_index_run(tenant_id: str, domain_id: str) -> str:
    with _connect() as conn:
        with conn.cursor() as cur:
            cur.execute(
                "SELECT index_run_id FROM index_run_active WHERE tenant_id = %s AND domain_id = %s",
                (tenant_id, domain_id),
            )
            row = cur.fetchone()
            if row:
                return str(row[0])
    run_id = create_index_run(tenant_id, domain_id)
    activate_index_run(tenant_id, domain_id, run_id)
    return run_id


def active_index_run_id(tenant_id: str, domain_id: str) -> str | None:
    with _connect() as conn:
        with conn.cursor() as cur:
            cur.execute(
                "SELECT index_run_id FROM index_run_active WHERE tenant_id = %s AND domain_id = %s",
                (tenant_id, domain_id),
            )
            row = cur.fetchone()
    return str(row[0]) if row else None


def upsert_index_document_state(
    run_id: str,
    document_id: str,
    indexed_version: int,
    content_sha256: str,
    chunk_count: int,
) -> None:
    with _connect() as conn:
        with conn.cursor() as cur:
            cur.execute(
                """
                INSERT INTO index_document_state
                    (index_run_id, document_id, indexed_version, content_sha256, chunk_count, indexed_at)
                VALUES (%s, %s, %s, %s, %s, NOW())
                ON CONFLICT (index_run_id, document_id) DO UPDATE SET
                    indexed_version = EXCLUDED.indexed_version,
                    content_sha256 = EXCLUDED.content_sha256,
                    chunk_count = EXCLUDED.chunk_count,
                    indexed_at = NOW()
                """,
                (run_id, document_id, indexed_version, content_sha256, chunk_count),
            )
        conn.commit()


def collection_suffix(tenant_id: str, domain_id: str) -> str:
    """Namespace suffix for disposable vector collections."""
    from rag.kb.index_collections import run_suffix

    run_id = active_index_run_id(tenant_id, domain_id)
    if not run_id:
        return ""
    return run_suffix(run_id)


def resolve_write_run_id(tenant_id: str, domain_id: str, explicit_run_id: str | None = None) -> str:
    """Run id for ingest writes: explicit building run or active (creating one if needed)."""
    if explicit_run_id:
        return explicit_run_id
    return ensure_active_index_run(tenant_id, domain_id)


def resolve_read_run_id(tenant_id: str, domain_id: str) -> str | None:
    """Active index run for retrieval; None → legacy flat index."""
    return active_index_run_id(tenant_id, domain_id)


def list_retired_run_ids(tenant_id: str, domain_id: str, *, keep_last: int = 1) -> list[str]:
    """Retired run ids eligible for GC (newest `keep_last` retired runs are kept)."""
    keep = max(0, int(keep_last))
    with _connect() as conn:
        with conn.cursor() as cur:
            cur.execute(
                """
                SELECT id FROM index_runs
                WHERE status = 'retired'
                  AND tenant_id IS NOT DISTINCT FROM NULLIF(%s, '')::text
                  AND domain_id IS NOT DISTINCT FROM NULLIF(%s, '')::text
                ORDER BY activated_at DESC NULLS LAST, created_at DESC
                OFFSET %s
                """,
                (tenant_id, domain_id, keep),
            )
            rows = cur.fetchall()
    return [str(row[0]) for row in rows]
